[R-package] Add docs on building and using precompiled binaries for the R package

jameslamb opened the pull request:
This PR comes out of this discussion: https://github.com/microsoft/LightGBM/pull/3188#issuecomment-664083054.

We're really close to getting the R package to CRAN, but I'm not sure how long it will take to fix  #3187 . To make life easier for users until the package gets to CRAN, starting with https://github.com/microsoft/LightGBM/releases/tag/v3.0.0rc1 I think we should start providing precompiled binaries for releases.

When #3283 is done, these binaries can be produced automatically and attached to the releases page. For now, I think it will be useful to document how to use them (for users) and how to build them (for maintainers and adventurous users).

StrikerRUS commented on R-package/README.md:
```diff
@@ -13,6 +13,36 @@ LightGBM R-package
+
+To install a binary for the R package:
+
+1. Choose a release from [the "Releases" page](https://github.com/microsoft/LightGBM/releases)
+2. Choose a file based on your operating system. Right-click it and choose "copy link address"
```
```suggestion
1. Choose a release from [the "Releases" page](https://github.com/microsoft/LightGBM/releases).
2. Choose a file based on your operating system. Right-click it and choose "copy link address".
```

StrikerRUS commented on R-package/README.md:
```diff
@@ -221,6 +251,73 @@ At build time, `configure.win` will be run and used to create a file `Makevars.w
+
+For more details, see ["Writing R Extensions"](https://cran.r-project.org/doc/manuals/r-release/R-exts.html#Building-binary-packages).
+
+Packages built like this will only work for the minor version of R used to build them. They may or may not work across different versions of operating systems
```
```suggestion
Packages built like this will only work for the minor version of R used to build them. They may or may not work across different versions of operating systems.
```

StrikerRUS commented on R-package/README.md:
```diff
@@ -221,6 +251,73 @@ At build time, `configure.win` will be run and used to create a file `Makevars.w
+
+**Mac**
+
+Binary produced: `lightgbm_${VERSION}-r40-macos.tgz`
```
```suggestion
Binary produced: `lightgbm_${VERSION}-r40-macos.tgz`.
```

StrikerRUS commented on R-package/README.md:
```diff
@@ -221,6 +251,73 @@ At build time, `configure.win` will be run and used to create a file `Makevars.w
+
+**Linux**
+
+Binary produced: `lightgbm_${VERSION}-r40-linux.tar.gz`
```
```suggestion
Binary produced: `lightgbm_${VERSION}-r40-linux.tar.gz`.
```

StrikerRUS commented on R-package/README.md:
````diff
@@ -221,6 +251,73 @@ At build time, `configure.win` will be run and used to create a file `Makevars.w
+        /bin/bash
+```
+
+From inside that container, the commands to create a precompiled binary are very similar
````
```suggestion
From inside that container, the commands to create a precompiled binary are very similar.
```

StrikerRUS commented on R-package/README.md:
```diff
@@ -221,6 +251,73 @@ At build time, `configure.win` will be run and used to create a file `Makevars.w
+
+**Windows**
+
+Binary produced: `lightgbm_${VERSION}.zip`
```
```suggestion
Binary produced: `lightgbm_${VERSION}.zip`.
```

StrikerRUS replied:
At release we have
```
lightgbm-${VERSION}.zip
        ^
```

![image](https://user-images.githubusercontent.com/25141164/89718606-87028f80-d9c8-11ea-9123-5be2230bd487.png)


Was it a typo?

StrikerRUS commented on R-package/README.md:
```diff
@@ -221,6 +251,73 @@ At build time, `configure.win` will be run and used to create a file `Makevars.w
+
+**Windows**
+
+Binary produced: `lightgbm_${VERSION}.zip`
```
At the Release page we have
```
lightgbm-${VERSION}.zip
        ^
```

![image](https://user-images.githubusercontent.com/25141164/89718606-87028f80-d9c8-11ea-9123-5be2230bd487.png)


Was it a typo?

jameslamb replied:
no I'll make them all `-` instead of `_` in the docs

StrikerRUS replied:
I listed some other inconsistent places in https://github.com/microsoft/LightGBM/pull/3285#pullrequestreview-463895208.

Also, I guess files should be renamed at Releases page:

![image](https://user-images.githubusercontent.com/25141164/89739630-f640b800-da8a-11ea-8676-3ccc54b0fd96.png)

jameslamb replied:
blegh sorry! Updated

jameslamb replied:
ok they are all consistent now

![image](https://user-images.githubusercontent.com/7608904/89853961-446fbd00-db58-11ea-8322-96c4783b0a6b.png)

StrikerRUS commented:
During reading this comment https://github.com/microsoft/LightGBM/issues/3187#issuecomment-670966615 I came to the thought that it might be useful to upload the source distribution to Releases page which will be then submitted to CRAN. WDYT @jameslamb ?

StrikerRUS commented on R-package/README.md:
````diff
@@ -13,6 +13,36 @@ LightGBM R-package
+This sample code installs versioon 3.0.0-1 of the R package on Mac.
+
+```r
+PKG_URL <- "https://github.com/microsoft/LightGBM/releases/download/v3.0.0rc1/lightgbm_3.0.0-1-r40-macos.tgz"
````
```suggestion
PKG_URL <- "https://github.com/microsoft/LightGBM/releases/download/v3.0.0rc1/lightgbm-3.0.0-1-r40-macos.tgz"
```

StrikerRUS commented on R-package/README.md:
```diff
@@ -221,6 +251,73 @@ At build time, `configure.win` will be run and used to create a file `Makevars.w
+R CMD INSTALL --build lightgbm_${LGB_VERSION}.tar.gz
+mv \
+    lightgbm_${LGB_VERSION}_R_*-linux-gnu.tar.gz \
+    lightgbm_${LGB_VERSION}-r40-linux.tgz
```
```suggestion
    lightgbm-${LGB_VERSION}-r40-linux.tgz
```

StrikerRUS commented on R-package/README.md:
```diff
@@ -221,6 +251,73 @@ At build time, `configure.win` will be run and used to create a file `Makevars.w
+
+**Windows**
+
+Binary produced: `lightgbm-${VERSION}.zip`.
```
```suggestion
Binary produced: `lightgbm-${VERSION}-r40-windows.zip`.
```

StrikerRUS commented on R-package/README.md:
```diff
@@ -221,6 +251,73 @@ At build time, `configure.win` will be run and used to create a file `Makevars.w
+R CMD INSTALL --build lightgbm_${LGB_VERSION}.tar.gz
+mv \
+    lightgbm_${LGB_VERSION}.tgz \
+    lightgbm-${LGB_VERSION}-r40-windows.tgz
```
```suggestion
    lightgbm-${LGB_VERSION}-r40-windows.zip
```

jameslamb commented:
> During reading this comment [#3187 (comment)](https://github.com/microsoft/LightGBM/issues/3187#issuecomment-670966615) I came to the thought that it might be useful to upload the source distribution to Releases page (automatically after resolving #3283 ) which will be then submitted to CRAN. WDYT @jameslamb ?

yes definitely!  I think that is a great idea and something we should be doing. I'll add it and a note in these docs.

And then we can add it automatically in future releases (#3283 )

jameslamb commented:
I just added a source distribution (built with `sh build-cran-package.sh`) to the 3.0.0rc1  release page!

I also added it to the list of artifacts that should be produced by #3283 

It can be installed like this from R.

```r
remotes::install_url(
    "https://github.com/microsoft/LightGBM/releases/download/v3.0.0rc1/lightgbm-3.0.0-1-cran.tar.gz"
)
```

jameslamb commented:
I just added notes on installing the CRAN package from the releases page, thanks for the nudge @StrikerRUS !

StrikerRUS commented:
@jameslamb 
> I also added it to the list of artifacts that should be produced by #3283

Thanks a lot!

I think after we put LightGBM on CRAN, we will have to produce only that artifact with source files and upload it to CRAN. All other artifacts with binaries should be uploaded reversely: from CRAN to Releases page (if needed) to have identical versions in both places and avoid errors like one just reported about glibc (https://github.com/microsoft/LightGBM/pull/3188#issuecomment-671088281). Please correct me if I'm wrong in my thoughts about further release pipeline.

jameslamb commented:
> @jameslamb
> 
> > I also added it to the list of artifacts that should be produced by #3283
> 
> Thanks a lot!
> 
> I think after we put LightGBM on CRAN, we will have to produce only that artifact with source files and upload it to CRAN. All other artifacts with binaries should be uploaded reversely: from CRAN to Releases page (if needed) to have identical versions in both places and avoid errors like one just reported about glibc ([#3188 (comment)](https://github.com/microsoft/LightGBM/pull/3188#issuecomment-671088281)). Please correct me if I'm wrong in my thoughts about further release pipeline.

Yes I agree with this. By the way, CRAN does not provide binaries for Linux (only Mac and Windows), so I think ours should be considered experimental.

StrikerRUS commented:
> By the way, CRAN does not provide binaries for Linux (only Mac and Windows), so I think ours should be considered experimental.

Hmm, then we should build them with our Docker we use for artifacts at Azure Pipelines to ensure maximum compatibility. Or even better with CentOS image.
https://github.com/pypa/manylinux/blob/master/docker/Dockerfile-x86_64

StrikerRUS commented on R-package/README.md:
```diff
@@ -13,6 +13,55 @@ LightGBM R-package
+To install this package on any operating system:
+
+1. Choose a release from [the "Releases" page](https://github.com/microsoft/LightGBM/releases)
+2. Look for the artifact with a name like `lgb-{VERSION}-r-cran.tar.gz`. Right-click it and choose "copy link address".
```
```suggestion
2. Look for the artifact with a name like `lightgbm-{VERSION}-r-cran.tar.gz`. Right-click it and choose "copy link address".
```

StrikerRUS commented on R-package/README.md:
```diff
@@ -221,6 +270,73 @@ At build time, `configure.win` will be run and used to create a file `Makevars.w
+
+**Linux**
+
+Binary produced: `lightgbm-${VERSION}-r40-linux.tar.gz`.
```
```suggestion
Binary produced: `lightgbm-${VERSION}-r40-linux.tgz`.
```

jameslamb commented:
thanks for keeping me consistent @StrikerRUS , sorry for all the errors.
